prboom2/src/dsda/utility/string_view.c

```diff
@@ -0,0 +1,125 @@
+//
+// Copyright(C) 2023 by Ryan Krafnick
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// DESCRIPTION:
+//	DSDA Utility String View
+//
+
+#include <string.h>
+
+#include "string_view.h"
+
+void dsda_InitStringView(dsda_string_view_t* sv, const char* string, size_t size) {
+  sv->string = string;
+  sv->size = size;
+}
+
+dboolean dsda_IsStringViewEmpty(const dsda_string_view_t* sv) {
+  return sv->size == 0;
+}
+
+void dsda_StringViewAtOffset(const dsda_string_view_t* sv, size_t offset,
+                               dsda_string_view_t* ofs) {
+  if (offset > sv->size) {
+    dsda_InitStringView(ofs, NULL, 0);
+    return;
+  }
+
+  ofs->string = sv->string + offset;
+  ofs->size = sv->size - offset;
+}
+
+dboolean dsda_SplitStringViewAfterChar(const dsda_string_view_t* sv, char c,
+                                       dsda_string_view_t* before,
+                                       dsda_string_view_t* after)
+{
+  const char* cur;
+  size_t size;
+
+  for (cur = sv->string, size = sv->size; size; ++cur, --size)
+    if (*cur == c) {
+      if (before)
+        dsda_InitStringView(before, sv->string, sv->size - size + 1);
+
+      if (after)
+        dsda_InitStringView(after, cur + 1, size - 1);
+
+      return true;
+    }
+
+  if (before)
+    dsda_InitStringView(before, sv->string, sv->size);
+
+  if (after)
+    dsda_InitStringView(after, NULL, 0);
+
+  return false;
+}
+
+dboolean dsda_SplitStringViewBeforeChar(const dsda_string_view_t* sv, char c,
+                                        dsda_string_view_t* before,
+                                        dsda_string_view_t* after)
+{
+  const char* cur;
+  size_t size;
+
+  for (cur = sv->string, size = sv->size; size; ++cur, --size)
+    if (*cur == c) {
+      if (before)
+        dsda_InitStringView(before, sv->string, sv->size - size);
+
+      if (after)
+        dsda_InitStringView(after, cur, size);
+
+      return true;
+    }
+
+  if (before)
+    dsda_InitStringView(before, sv->string, sv->size);
+
+  if (after)
+    dsda_InitStringView(after, NULL, 0);
+
+  return false;
+}
+
+dboolean dsda_AdvanceStringViewToNextLine(dsda_string_view_t* sv, dsda_string_view_t* line) {
```

Comment:
What do you think about switching this to something like `dsda_GetStringViewLine` that sets line to the _current_ line and advances the string view forward (so treating it more like reading from a stream)?

Reply:
That's what this function does.  `line` is set to the the line starting at the current position of `sv`, and `sv` is advanced past it.

Reply:
Ah ya, now I see the line is in the before slot. I think the comment before wasn't clear. Maybe the name should be indeed be `dsda_GetStringViewLine` then.

Reply:
Having `Next` in the name emphasizes the iterator-like behavior, but yeah.